flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java

```diff
@@ -262,36 +261,92 @@ public void grantLeadership(UUID leaderSessionID) {
+                                    } else {
+                                        createNewJobMasterServiceProcessIfValidLeader(
+                                                leaderSessionId);
+                                    }
+                                    return null;
+                                });
         handleAsyncOperationError(sequentialOperation, "Could not start the job manager.");
     }
 
-    @GuardedBy("lock")
-    private void verifyJobSchedulingStatusAndCreateJobMasterServiceProcess(UUID leaderSessionId)
-            throws FlinkException {
-        try {
-            if (jobResultStore.hasJobResultEntry(getJobID())) {
-                jobAlreadyDone(leaderSessionId);
+    private <T> CompletableFuture<T> supplyAsyncIfValidLeader(
+            UUID expectedLeaderId,
+            Supplier<CompletableFuture<T>> supplier,
+            Supplier<CompletableFuture<T>> noLeaderFallback) {
+        final CompletableFuture<T> resultFuture = new CompletableFuture<>();
+        runIfValidLeader(
+                expectedLeaderId,
+                () -> FutureUtils.forward(supplier.get(), resultFuture),
+                () -> FutureUtils.forward(noLeaderFallback.get(), resultFuture));
+
+        return resultFuture;
+    }
+
+    private void handleJobAlreadyDoneIfValidLeader(UUID leaderSessionId) {
+        runIfValidLeader(
+                leaderSessionId,
+                () -> {
+                    jobAlreadyDone(leaderSessionId);
+                },
+                () -> printLogIfNotValidLeader("check completed job", leaderSessionId));
+    }
+
+    private void createNewJobMasterServiceProcessIfValidLeader(UUID leaderSessionId) {
+        runIfValidLeader(
+                leaderSessionId,
+                () -> {
+                    ThrowingRunnable.unchecked(
+                                    () -> createNewJobMasterServiceProcess(leaderSessionId))
+                            .run();
+                },
+                () ->
+                        printLogIfNotValidLeader(
+                                "create new job master service process", leaderSessionId));
+    }
+
+    private void runIfValidLeader(
+            UUID expectedLeaderId, Runnable action, Runnable noLeaderFallback) {
+        synchronized (lock) {
+            if (isValidLeader(expectedLeaderId)) {
+                action.run();
             } else {
-                createNewJobMasterServiceProcess(leaderSessionId);
+                noLeaderFallback.run();
             }
-        } catch (IOException e) {
-            throw new FlinkException(
-                    String.format(
-                            "Could not retrieve the job scheduling status for job %s.", getJobID()),
-                    e);
         }
     }
 
+    private CompletableFuture<Boolean> printLogIfNotValidLeader(
+            String actionDescription, UUID leaderSessionId) {
+        LOG.trace(
```

Comment:
a bit out-of-scope, but: I feel like these log statement should be in debug level. :thinking: WDYT? But we should fix it in a separate hotfix commit if you agree.

Reply:
Thanks for the suggestion! I've followed the 2nd solution.